mlflow/store/artifact/databricks_artifact_repo.py

```diff
@@ -196,6 +218,76 @@ def _get_write_credential_infos(self, remote_file_paths):
             GetCredentialsForWrite, self.run_id, run_relative_remote_paths
         )
 
+    def download_trace(self, trace_id):
+        """
+        Downloads the trace data for the specified trace ID.
+
+        Args:
+            trace_id: The trace ID for which to download trace data.
+
+        Returns:
+            A dictionary containing the trace data.
+        """
+        cred = self._call_endpoint(
+            DatabricksMlflowArtifactsService,
+            GetCredentialsForTraceDataDownload,
+            None,
+            {"trace_id": trace_id},
+        )
+        signed_uri = cred.credential_info.signed_uri
+        headers = self._extract_headers_from_credentials(cred.credential_info.headers)
+        with tempfile.TemporaryDirectory() as temp_dir:
+            local_path = Path(temp_dir, "traces.json")
+            download_file_using_http_uri(
+                signed_uri,
+                local_path,
+                headers=headers,
+            )
+            with local_path.open("r") as f:
+                return json.load(f)
+
+    def upload_trace(self, trace: Trace):
+        """
+        Uploads the trace data for the specified trace ID.
+
+        Args:
+            trace: The trace data to upload.
+
+        Returns:
+            None
+        """
+        cred = self._call_endpoint(
+            DatabricksMlflowArtifactsService,
+            GetCredentialsForTraceDataUpload,
+            None,
+            {"trace_id": trace.trace_info.trace_id},
+        )
+        with tempfile.TemporaryDirectory() as temp_dir:
+            temp_file = Path(temp_dir, "traces.json")
+            with temp_file.open("w") as f:
+                f.write(trace.to_json())
+
+            if cred.credential_info.type == ArtifactCredentialType.AZURE_ADLS_GEN2_SAS_URI:
+                signed_uri = cred.credential_info.signed_uri
+                headers = self._extract_headers_from_credentials(cred.credential_info.headers)
+                size = temp_file.stat().st_size
+                put_adls_file_creation(sas_url=signed_uri, headers=headers)
+                patch_adls_file_upload(
+                    sas_url=signed_uri,
+                    local_file=temp_file,
+                    start_byte=0,
+                    size=size,
+                    position=0,
+                    headers=headers,
+                    is_single=True,
+                )
```

Comment:
Should we be reusing `_azure_adls_gen2_upload_file` instead? One thing that's missing from the current impl is retries. `_azure_adls_gen2_upload_file` uses `_retryable_adls_function`

Reply:
Do we want to retry? `_signed_url_upload_file` doesn't do a retry.

Reply:
`_retryable_adls_function` uses `self._get_write_credential_infos` to generate a new credential. That needs to be changed.

Reply:
For ADLS, we always use multipart-upload. We don't want to retry the upload just because one chunk upload fails. I think that's the reason.

Reply:
Makes sense. As discussed offline, we can add retries later if necessary